### PR TITLE
Introduce magnetic field caching in vertexing and remove refitAfterBadVertex option

### PR DESCRIPTION
**Pull request by baschlag**
This PR introduces bfield caching in the vertexing and removes the ```refitAfterBadVertex``` option in the AMVF as it evidently worsens the physics performance of the finder.

### CODE REVIEW

**Comment by baschlag** [comment]
FYI: @paulgessinger

**Comment by baschlag** [comment]
> General remarks:
> If it's unrelated (state and refitAfterBadVertex) make it 2 PRs, even if they are small each, it's just easier if every change is for the same thing :)
> I wouldn't have put the magFieldCache into a state as long as it's just one single variable but that's a question of taste
> Bonus question: did you understand why your refitAfterBadVertex did not yield the same results as was expected?

1. So first I had to remove the refitAfterBadVertex option in this PR together with the other changes as it seems to have thrown errors when copying over the fitterState. That's why I had both of them in one PR together. Maybe this has been resolved and I could split them up.. I will check this.
2. So you would just pass the fieldCache by reference instead of a state holding the fieldCache?
3. Yes. If a bad vertex gets removed and I just return to the old fitter state (which saves some time), I don't have the additional fits of all vertices which worsens the resolution compared to the case where I do have additional vertex fits inbetween.

File: Core/include/Acts/Vertexing/HelicalTrackLinearizer.ipp [hunk 1]
```diff
@@ -62,7 +62,8 @@ Acts::Result<Acts::LinearizedTrack> Acts::
   Vector3D momentumAtPCA(phiV, th, qOvP);
 
   // get B-field z-component at current position
-  double Bz = m_cfg.bField.getField(VectorHelpers::position(positionAtPCA))[eZ];
+  double Bz = m_cfg.bField.getField(VectorHelpers::position(positionAtPCA),
+                                    state.fieldCache)[eZ];
```
[the remark this answers is not in the document]
Comment (baschlag) [reply Core/include/Acts/Vertexing/HelicalTrackLinearizer.ipp:66]: The state constructors have no default constructor but only one taking the magneticFieldContext. In that way it's ensured that it's always initialized.

**Comment by robertlangenberg** [comment]
>     1. So first I had to remove the refitAfterBadVertex option in this PR together with the other changes as it seems to have thrown errors when copying over the fitterState. That's why I had both of them in one PR together. Maybe this has been resolved and I could split them up.. I will check this.

No need, just for next time. If you had your reasons, never mind this comment :)

>     2. So you would just pass the fieldCache by reference instead of a state holding the fieldCache?

Yes but I consider this a question of personal taste. It's fine to leave it if you prefer it like this.

>     3. Yes. If a bad vertex gets removed and I just return to the old fitter state (which saves some time), I don't have the additional fits of all vertices which worsens the resolution compared to the case where I do have additional vertex fits inbetween.

Ok, thanks!

**Comment by msmk0** [comment]
@baschlag This PR still targets the master branch. We can only merge this after we fix the CI issue and tag v0.28.0. If you want to get this into v0.27.1 you should open a new PR.

**Comment by baschlag** [comment]
> @baschlag This PR still targets the master branch. We can only merge this after we fix the CI issue and tag v0.28.0. If you want to get this into v0.27.1 you should open a new PR.

Hey @msmk0, yes I know. This one is supposed to go into master. I will open a PR soon targeting the release v0.27 branch for backporting into v0.27.1 for the Athena integration.

**Comment by msmk0** [comment]
@baschlag Could you merge in the latest master branch. I will merge it in afterwards and then we can tag 0.28.0.

**Comment by baschlag** [comment]
> @baschlag Could you merge in the latest master branch. I will merge it in afterwards and then we can tag 0.28.0.

Done. But we still keep the v0.27.1 backport PR?

**Comment by msmk0** [comment]
> Done. But we still keep the v0.27.1 backport PR?

Yes.

**Comment by msmk0** [comment]
@robertlangenberg @paulgessinger Could you re-approve?

**Pull request merged**